File: backend/vyos_mappers/vpn/ipsec.py
"""
IPsec VPN Command Mapper

Handles IPsec VPN commands for site-to-site tunnels.
"""

from typing import List, Dict, Any, Optional
from ..base import BaseFeatureMapper


class IPsecMapper(BaseFeatureMapper):
    """IPsec VPN mapper with all IPsec operations"""

    def __init__(self, version: str):
        """Initialize with VyOS version."""
        super().__init__(version)

    # ========================================================================
    # IKE Group Commands
    # ========================================================================

    def get_ike_group(self, name: str) -> List[str]:
        """Get command path for creating an IKE group."""
        return ["vpn", "ipsec", "ike-group", name]

    def get_ike_group_proposal(self, name: str, proposal: str) -> List[str]:
        """Get command path for IKE group proposal."""
        return ["vpn", "ipsec", "ike-group", name, "proposal", proposal]

    def get_ike_group_proposal_dh_group(self, name: str, proposal: str, dh_group: str) -> List[str]:
        """Get command path for IKE proposal DH group."""
        return ["vpn", "ipsec", "ike-group", name, "proposal", proposal, "dh-group", dh_group]

    def get_ike_group_proposal_encryption(self, name: str, proposal: str, encryption: str) -> List[str]:
        """Get command path for IKE proposal encryption algorithm."""
        return ["vpn", "ipsec", "ike-group", name, "proposal", proposal, "encryption", encryption]

    def get_ike_group_proposal_hash(self, name: str, proposal: str, hash_algo: str) -> List[str]:
        """Get command path for IKE proposal hash algorithm."""
        return ["vpn", "ipsec", "ike-group", name, "proposal", proposal, "hash", hash_algo]

    def get_ike_group_key_exchange(self, name: str, version: str) -> List[str]:
        """Get command path for IKE key exchange version (ikev1, ikev2)."""
        return ["vpn", "ipsec", "ike-group", name, "key-exchange", version]

    def get_ike_group_lifetime(self, name: str, lifetime: str) -> List[str]:
        """Get command path for IKE group lifetime (seconds)."""
        return ["vpn", "ipsec", "ike-group", name, "lifetime", lifetime]

    def get_ike_group_dpd_action(self, name: str, action: str) -> List[str]:
        """Get command path for DPD action (clear, hold, restart)."""
        return ["vpn", "ipsec", "ike-group", name, "dead-peer-detection", "action", action]

    def get_ike_group_dpd_interval(self, name: str, interval: str) -> List[str]:
        """Get command path for DPD interval."""
        return ["vpn", "ipsec", "ike-group", name, "dead-peer-detection", "interval", interval]

    def get_ike_group_dpd_timeout(self, name: str, timeout: str) -> List[str]:
        """Get command path for DPD timeout."""
        return ["vpn", "ipsec", "ike-group", name, "dead-peer-detection", "timeout", timeout]

    def get_ike_group_close_action(self, name: str, action: str) -> List[str]:
        """Get command path for close action (none, clear, hold, restart)."""
        return ["vpn", "ipsec", "ike-group", name, "close-action", action]

    def get_ike_group_ikev2_reauth(self, name: str) -> List[str]:
        """Get command path for IKEv2 reauth (yes/no)."""
        return ["vpn", "ipsec", "ike-group", name, "ikev2-reauth"]

    def get_ike_group_mode(self, name: str, mode: str) -> List[str]:
        """Get command path for IKE mode (main, aggressive)."""
        return ["vpn", "ipsec", "ike-group", name, "mode", mode]

    # ========================================================================
    # ESP Group Commands
    # ========================================================================

    def get_esp_group(self, name: str) -> List[str]:
        """Get command path for creating an ESP group."""
        return ["vpn", "ipsec", "esp-group", name]

    def get_esp_group_proposal(self, name: str, proposal: str) -> List[str]:
        """Get command path for ESP group proposal."""
        return ["vpn", "ipsec", "esp-group", name, "proposal", proposal]

    def get_esp_group_proposal_encryption(self, name: str, proposal: str, encryption: str) -> List[str]:
        """Get command path for ESP proposal encryption algorithm."""
        return ["vpn", "ipsec", "esp-group", name, "proposal", proposal, "encryption", encryption]

    def get_esp_group_proposal_hash(self, name: str, proposal: str, hash_algo: str) -> List[str]:
        """Get command path for ESP proposal hash algorithm."""
        return ["vpn", "ipsec", "esp-group", name, "proposal", proposal, "hash", hash_algo]

    def get_esp_group_lifetime(self, name: str, lifetime: str) -> List[str]:
        """Get command path for ESP group lifetime (seconds)."""
        return ["vpn", "ipsec", "esp-group", name, "lifetime", lifetime]

    def get_esp_group_pfs(self, name: str, pfs: str) -> List[str]:
        """Get command path for ESP group PFS (dh-group1, dh-group2, etc.)."""
        return ["vpn", "ipsec", "esp-group", name, "pfs", pfs]

    def get_esp_group_mode(self, name: str, mode: str) -> List[str]:
        """Get command path for ESP group mode (tunnel, transport)."""
        return ["vpn", "ipsec", "esp-group", name, "mode", mode]

    def get_esp_group_compression(self, name: str) -> List[str]:
        """Get command path for ESP group compression."""
        return ["vpn", "ipsec", "esp-group", name, "compression"]

    # ========================================================================
    # Site-to-Site Peer Commands
    # ========================================================================

    def get_site_to_site_peer(self, peer: str) -> List[str]:
        """Get command path for creating a site-to-site peer."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer]

    def get_peer_authentication_mode(self, peer: str, mode: str) -> List[str]:
        """Get command path for peer auth mode (pre-shared-secret, x509)."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "mode", mode]

    def get_peer_authentication_psk(self, peer: str, psk: str) -> List[str]:
        """Get command path for peer pre-shared-key."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "pre-shared-secret", psk]

    def get_peer_authentication_local_id(self, peer: str, local_id: str) -> List[str]:
        """Get command path for local ID."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "local-id", local_id]

    def get_peer_authentication_remote_id(self, peer: str, remote_id: str) -> List[str]:
        """Get command path for remote ID."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "remote-id", remote_id]

    def get_peer_authentication_x509_ca_cert(self, peer: str, cert: str) -> List[str]:
        """Get command path for X.509 CA certificate."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "x509", "ca-certificate", cert]

    def get_peer_authentication_x509_cert(self, peer: str, cert: str) -> List[str]:
        """Get command path for X.509 certificate."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "authentication", "x509", "certificate", cert]

    def get_peer_connection_type(self, peer: str, conn_type: str) -> List[str]:
        """Get command path for connection type (initiate, respond)."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "connection-type", conn_type]

    def get_peer_default_esp_group(self, peer: str, esp_group: str) -> List[str]:
        """Get command path for peer default ESP group."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "default-esp-group", esp_group]

    def get_peer_ike_group(self, peer: str, ike_group: str) -> List[str]:
        """Get command path for peer IKE group."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "ike-group", ike_group]

    def get_peer_local_address(self, peer: str, address: str) -> List[str]:
        """Get command path for peer local address."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "local-address", address]

    def get_peer_description(self, peer: str, description: str) -> List[str]:
        """Get command path for peer description."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "description", description]

    def get_peer_description_path(self, peer: str) -> List[str]:
        """Get command path for peer description (for deletion)."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "description"]

    def get_peer_disable(self, peer: str) -> List[str]:
        """Get command path for disabling a peer."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "disable"]

    def get_peer_dhcp_interface(self, peer: str, interface: str) -> List[str]:
        """Get command path for DHCP interface (dynamic local address)."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "dhcp-interface", interface]

    def get_peer_vti_bind(self, peer: str, vti: str) -> List[str]:
        """Get command path for VTI binding."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "vti", "bind", vti]

    def get_peer_vti_esp_group(self, peer: str, esp_group: str) -> List[str]:
        """Get command path for VTI ESP group."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "vti", "esp-group", esp_group]

    # Tunnel configuration
    def get_peer_tunnel(self, peer: str, tunnel_id: str) -> List[str]:
        """Get command path for peer tunnel."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id]

    def get_peer_tunnel_esp_group(self, peer: str, tunnel_id: str, esp_group: str) -> List[str]:
        """Get command path for tunnel ESP group."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id, "esp-group", esp_group]

    def get_peer_tunnel_local_prefix(self, peer: str, tunnel_id: str, prefix: str) -> List[str]:
        """Get command path for tunnel local prefix."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id, "local", "prefix", prefix]

    def get_peer_tunnel_remote_prefix(self, peer: str, tunnel_id: str, prefix: str) -> List[str]:
        """Get command path for tunnel remote prefix."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id, "remote", "prefix", prefix]

    def get_peer_tunnel_protocol(self, peer: str, tunnel_id: str, protocol: str) -> List[str]:
        """Get command path for tunnel protocol filter (gre, all, etc.)."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id, "protocol", protocol]

    def get_peer_tunnel_disable(self, peer: str, tunnel_id: str) -> List[str]:
        """Get command path for disabling a tunnel."""
        return ["vpn", "ipsec", "site-to-site", "peer", peer, "tunnel", tunnel_id, "disable"]

    # ========================================================================
    # IPsec Interface Commands
    # ========================================================================

    def get_ipsec_interface(self, interface: str) -> List[str]:
        """Get command path for IPsec interface."""
        return ["vpn", "ipsec", "interface", interface]

    # ========================================================================
    # IPsec Options Commands
    # ========================================================================

    def get_options_disable_route_autoinstall(self) -> List[str]:
        """Get command path for disabling route auto-install."""
        return ["vpn", "ipsec", "options", "disable-route-autoinstall"]

    def get_options_flexvpn(self) -> List[str]:
        """Get command path for enabling FlexVPN."""
        return ["vpn", "ipsec", "options", "flexvpn"]

    def get_options_virtual_ip(self, ip: str) -> List[str]:
        """Get command path for virtual IP."""
        return ["vpn", "ipsec", "options", "virtual-ip", ip]

    # ========================================================================
    # Config Parsing Methods (for READ operations)
    # ========================================================================

    def parse_ike_groups(self, config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse IKE groups from config."""
        ike_groups = []
        ike_config = config.get("ike-group", {})

        for name, group_config in ike_config.items():
            if not isinstance(group_config, dict):
                continue

            # Parse proposals
            proposals = []
            prop_config = group_config.get("proposal", {})
            for prop_id, prop_data in prop_config.items() if isinstance(prop_config, dict) else []:
                if isinstance(prop_data, dict):
                    proposals.append({
                        "id": prop_id,
                        "dh_group": prop_data.get("dh-group"),
                        "encryption": prop_data.get("encryption"),
                        "hash": prop_data.get("hash"),
                    })

            # Parse DPD
            dpd = None
            dpd_config = group_config.get("dead-peer-detection", {})
            if isinstance(dpd_config, dict) and dpd_config:
                dpd = {
                    "action": dpd_config.get("action"),
                    "interval": dpd_config.get("interval"),
                    "timeout": dpd_config.get("timeout"),
                }

            ike_groups.append({
                "name": name,
                "key_exchange": group_config.get("key-exchange"),
                "lifetime": group_config.get("lifetime"),
                "proposals": proposals,
                "dead_peer_detection": dpd,
                "close_action": group_config.get("close-action"),
                "ikev2_reauth": "ikev2-reauth" in group_config,
                "mode": group_config.get("mode"),
            })

        return ike_groups

    def parse_esp_groups(self, config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse ESP groups from config."""
        esp_groups = []
        esp_config = config.get("esp-group", {})

        for name, group_config in esp_config.items():
            if not isinstance(group_config, dict):
                continue

            # Parse proposals
            proposals = []
            prop_config = group_config.get("proposal", {})
            for prop_id, prop_data in prop_config.items() if isinstance(prop_config, dict) else []:
                if isinstance(prop_data, dict):
                    proposals.append({
                        "id": prop_id,
                        "encryption": prop_data.get("encryption"),
                        "hash": prop_data.get("hash"),
                    })

            esp_groups.append({
                "name": name,
                "lifetime": group_config.get("lifetime"),
                "pfs": group_config.get("pfs"),
                "mode": group_config.get("mode", "tunnel"),
                "compression": "compression" in group_config,
                "proposals": proposals,
            })

        return esp_groups

    def parse_site_to_site_peers(self, config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse site-to-site peers from config."""
        peers = []
        sts_config = config.get("site-to-site", {}).get("peer", {})

        for peer_addr, peer_config in sts_config.items():
            if not isinstance(peer_config, dict):
                continue

            # Parse authentication
            auth = {}
            auth_config = peer_config.get("authentication", {})
            if isinstance(auth_config, dict):
                auth = {
                    "mode": auth_config.get("mode"),
                    "pre_shared_secret": auth_config.get("pre-shared-secret"),
                    "local_id": auth_config.get("local-id"),
                    "remote_id": auth_config.get("remote-id"),
                }
                # X.509 certs
                x509 = auth_config.get("x509", {})
                if isinstance(x509, dict):
                    auth["x509"] = {
                        "ca_certificate": x509.get("ca-certificate"),
                        "certificate": x509.get("certificate"),
                    }

            # Parse tunnels
            tunnels = []
            tunnel_config = peer_config.get("tunnel", {})
            for tunnel_id, tunnel_data in tunnel_config.items() if isinstance(tunnel_config, dict) else []:
                if isinstance(tunnel_data, dict):
                    local_prefix = None
                    remote_prefix = None
                    local_config = tunnel_data.get("local", {})
                    remote_config = tunnel_data.get("remote", {})
                    if isinstance(local_config, dict):
                        local_prefix = local_config.get("prefix")
                    if isinstance(remote_config, dict):
                        remote_prefix = remote_config.get("prefix")

                    tunnels.append({
                        "id": tunnel_id,
                        "esp_group": tunnel_data.get("esp-group"),
                        "local_prefix": local_prefix,
                        "remote_prefix": remote_prefix,
                        "protocol": tunnel_data.get("protocol"),
                        "disable": "disable" in tunnel_data,
                    })

            # Parse VTI
            vti = None
            vti_config = peer_config.get("vti", {})
            if isinstance(vti_config, dict) and vti_config:
                vti = {
                    "bind": vti_config.get("bind"),
                    "esp_group": vti_config.get("esp-group"),
                }

            peers.append({
                "address": peer_addr,
                "authentication": auth,
                "connection_type": peer_config.get("connection-type"),
                "default_esp_group": peer_config.get("default-esp-group"),
                "ike_group": peer_config.get("ike-group"),
                "local_address": peer_config.get("local-address"),
                "description": peer_config.get("description"),
                "disable": "disable" in peer_config,
                "dhcp_interface": peer_config.get("dhcp-interface"),
                "vti": vti,
                "tunnels": tunnels,
            })

        return peers

    def parse_interfaces(self, config: Dict[str, Any]) -> List[str]:
        """Parse IPsec interfaces from config."""
        interfaces = config.get("interface", [])
        if isinstance(interfaces, str):
            return [interfaces]
        elif isinstance(interfaces, list):
            return interfaces
        return []

    def parse_options(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """Parse IPsec options from config."""
        options_config = config.get("options", {})
        if not isinstance(options_config, dict):
            return {}

        virtual_ips = options_config.get("virtual-ip", [])
        if isinstance(virtual_ips, str):
            virtual_ips = [virtual_ips]

        return {
            "disable_route_autoinstall": "disable-route-autoinstall" in options_config,
            "flexvpn": "flexvpn" in options_config,
            "virtual_ips": virtual_ips if isinstance(virtual_ips, list) else [],
        }

    def parse_full_config(self, full_config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Parse full IPsec configuration from VyOS.

        Args:
            full_config: Full VyOS config dictionary

        Returns:
            Parsed IPsec configuration
        """
        ipsec_config = full_config.get("vpn", {}).get("ipsec", {})

        if not ipsec_config:
            return {
                "configured": False,
                "ike_groups": [],
                "esp_groups": [],
                "peers": [],
                "interfaces": [],
                "options": {},
            }

        return {
            "configured": True,
            "ike_groups": self.parse_ike_groups(ipsec_config),
            "esp_groups": self.parse_esp_groups(ipsec_config),
            "peers": self.parse_site_to_site_peers(ipsec_config),
            "interfaces": self.parse_interfaces(ipsec_config),
            "options": self.parse_options(ipsec_config),
        }

File: frontend/src/lib/api/ipsec.ts
/**
 * IPsec VPN API Service
 * Handles all IPsec VPN related API operations
 */

import { apiClient } from "./client";

// ============================================================================
// Types
// ============================================================================

export interface IKEProposal {
  id: string;
  dh_group?: string;
  encryption?: string;
  hash?: string;
}

export interface DeadPeerDetection {
  action?: string;
  interval?: string;
  timeout?: string;
}

export interface IKEGroup {
  name: string;
  key_exchange?: string;
  lifetime?: string;
  proposals: IKEProposal[];
  dead_peer_detection?: DeadPeerDetection;
  close_action?: string;
  ikev2_reauth: boolean;
  mode?: string;
}

export interface ESPProposal {
  id: string;
  encryption?: string;
  hash?: string;
}

export interface ESPGroup {
  name: string;
  lifetime?: string;
  pfs?: string;
  mode: string;
  compression: boolean;
  proposals: ESPProposal[];
}

export interface PeerAuthentication {
  mode?: string;
  pre_shared_secret?: string;
  local_id?: string;
  remote_id?: string;
  x509?: {
    ca_certificate?: string;
    certificate?: string;
  };
}

export interface PeerTunnel {
  id: string;
  esp_group?: string;
  local_prefix?: string;
  remote_prefix?: string;
  protocol?: string;
  disable: boolean;
}

export interface PeerVTI {
  bind?: string;
  esp_group?: string;
}

export interface SiteToSitePeer {
  address: string;
  authentication?: PeerAuthentication;
  connection_type?: string;
  default_esp_group?: string;
  ike_group?: string;
  local_address?: string;
  description?: string;
  disable: boolean;
  dhcp_interface?: string;
  vti?: PeerVTI;
  tunnels: PeerTunnel[];
}

export interface IPsecOptions {
  disable_route_autoinstall: boolean;
  flexvpn: boolean;
  virtual_ips: string[];
}

export interface IPsecConfig {
  configured: boolean;
  ike_groups: IKEGroup[];
  esp_groups: ESPGroup[];
  peers: SiteToSitePeer[];
  interfaces: string[];
  options?: IPsecOptions;
}

export interface IPsecCapabilities {
  encryptions: { value: string; label: string; description: string }[];
  hashes: { value: string; label: string; description: string }[];
  dh_groups: { value: string; label: string; description: string }[];
  pfs_options: { value: string; label: string; description: string }[];
  key_exchanges: { value: string; label: string; description: string }[];
  connection_types: { value: string; label: string; description: string }[];
  dpd_actions: { value: string; label: string; description: string }[];
  auth_modes: { value: string; label: string; description: string }[];
  esp_modes: { value: string; label: string; description: string }[];
  default_lifetimes: { ike: number; esp: number };
  version: string;
}

export interface IPsecOperation {
  op: string;
  name?: string;
  value?: string | number;
  peer?: string;
  proposal?: string;
  tunnel?: string;
  interface?: string;
  dh_group?: string;
  encryption?: string;
  hash?: string;
  action?: string;
  interval?: number;
  timeout?: number;
  esp_group?: string;
  local_prefix?: string;
  remote_prefix?: string;
  protocol?: string;
  bind?: string;
}

export interface IPsecBatchRequest {
  operations: IPsecOperation[];
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown>;
  error?: string;
}

// ============================================================================
// Service
// ============================================================================

class IPsecService {
  /**
   * Get IPsec configuration
   */
  async getConfig(): Promise<IPsecConfig> {
    return apiClient.get<IPsecConfig>("/vyos/ipsec/config");
  }

  /**
   * Get IPsec capabilities for the connected VyOS version
   */
  async getCapabilities(): Promise<IPsecCapabilities> {
    return apiClient.get<IPsecCapabilities>("/vyos/ipsec/capabilities");
  }

  /**
   * Get IPsec tunnel status
   */
  async getStatus(): Promise<{ success: boolean; data: Record<string, unknown> }> {
    return apiClient.get("/vyos/ipsec/status");
  }

  /**
   * Get IPsec connections
   */
  async getConnections(): Promise<{ success: boolean; data: Record<string, unknown> }> {
    return apiClient.get("/vyos/ipsec/connections");
  }

  /**
   * Configure IPsec using batch operations
   */
  async configureBatch(request: IPsecBatchRequest): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/ipsec/batch", request);
  }

  // =========================================================================
  // IKE Group Helper Methods
  // =========================================================================

  /**
   * Create an IKE group
   */
  async createIKEGroup(
    name: string,
    options?: {
      keyExchange?: string;
      lifetime?: number;
      proposal?: {
        id: string;
        dhGroup?: string;
        encryption?: string;
        hash?: string;
      };
      dpd?: {
        action?: string;
        interval?: number;
        timeout?: number;
      };
    }
  ): Promise<VyOSResponse> {
    const operations: IPsecOperation[] = [{ op: "create_ike_group", name }];

    if (options?.keyExchange) {
      operations.push({ op: "set_ike_group_key_exchange", name, value: options.keyExchange });
    }

    if (options?.lifetime) {
      operations.push({ op: "set_ike_group_lifetime", name, value: options.lifetime });
    }

    if (options?.proposal) {
      operations.push({
        op: "add_ike_group_proposal",
        name,
        proposal: options.proposal.id,
        dh_group: options.proposal.dhGroup,
        encryption: options.proposal.encryption,
        hash: options.proposal.hash,
      });
    }

    if (options?.dpd) {
      operations.push({
        op: "set_ike_group_dpd",
        name,
        action: options.dpd.action,
        interval: options.dpd.interval,
        timeout: options.dpd.timeout,
      });
    }

    return this.configureBatch({ operations });
  }

  /**
   * Delete an IKE group
   */
  async deleteIKEGroup(name: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "delete_ike_group", name }],
    });
  }

  /**
   * Add a proposal to an IKE group
   */
  async addIKEProposal(
    groupName: string,
    proposalId: string,
    options: {
      dhGroup?: string;
      encryption?: string;
      hash?: string;
    }
  ): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{
        op: "add_ike_group_proposal",
        name: groupName,
        proposal: proposalId,
        dh_group: options.dhGroup,
        encryption: options.encryption,
        hash: options.hash,
      }],
    });
  }

  // =========================================================================
  // ESP Group Helper Methods
  // =========================================================================

  /**
   * Create an ESP group
   */
  async createESPGroup(
    name: string,
    options?: {
      lifetime?: number;
      pfs?: string;
      mode?: string;
      proposal?: {
        id: string;
        encryption?: string;
        hash?: string;
      };
    }
  ): Promise<VyOSResponse> {
    const operations: IPsecOperation[] = [{ op: "create_esp_group", name }];

    if (options?.lifetime) {
      operations.push({ op: "set_esp_group_lifetime", name, value: options.lifetime });
    }

    if (options?.pfs) {
      operations.push({ op: "set_esp_group_pfs", name, value: options.pfs });
    }

    if (options?.mode) {
      operations.push({ op: "set_esp_group_mode", name, value: options.mode });
    }

    if (options?.proposal) {
      operations.push({
        op: "add_esp_group_proposal",
        name,
        proposal: options.proposal.id,
        encryption: options.proposal.encryption,
        hash: options.proposal.hash,
      });
    }

    return this.configureBatch({ operations });
  }

  /**
   * Delete an ESP group
   */
  async deleteESPGroup(name: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "delete_esp_group", name }],
    });
  }

  /**
   * Add a proposal to an ESP group
   */
  async addESPProposal(
    groupName: string,
    proposalId: string,
    options: {
      encryption?: string;
      hash?: string;
    }
  ): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{
        op: "add_esp_group_proposal",
        name: groupName,
        proposal: proposalId,
        encryption: options.encryption,
        hash: options.hash,
      }],
    });
  }

  // =========================================================================
  // Site-to-Site Peer Helper Methods
  // =========================================================================

  /**
   * Create a site-to-site peer
   */
  async createPeer(
    peerAddress: string,
    options: {
      ikeGroup: string;
      localAddress?: string;
      authMode?: string;
      preSharedKey?: string;
      localId?: string;
      remoteId?: string;
      connectionType?: string;
      defaultEspGroup?: string;
      description?: string;
    }
  ): Promise<VyOSResponse> {
    const operations: IPsecOperation[] = [
      { op: "create_peer", peer: peerAddress },
      { op: "set_peer_ike_group", peer: peerAddress, value: options.ikeGroup },
    ];

    if (options.localAddress) {
      operations.push({ op: "set_peer_local_address", peer: peerAddress, value: options.localAddress });
    }

    if (options.authMode) {
      operations.push({ op: "set_peer_auth_mode", peer: peerAddress, value: options.authMode });
    }

    if (options.preSharedKey) {
      operations.push({ op: "set_peer_auth_psk", peer: peerAddress, value: options.preSharedKey });
    }

    if (options.localId) {
      operations.push({ op: "set_peer_auth_local_id", peer: peerAddress, value: options.localId });
    }

    if (options.remoteId) {
      operations.push({ op: "set_peer_auth_remote_id", peer: peerAddress, value: options.remoteId });
    }

    if (options.connectionType) {
      operations.push({ op: "set_peer_connection_type", peer: peerAddress, value: options.connectionType });
    }

    if (options.defaultEspGroup) {
      operations.push({ op: "set_peer_default_esp_group", peer: peerAddress, value: options.defaultEspGroup });
    }

    if (options.description) {
      operations.push({ op: "set_peer_description", peer: peerAddress, value: options.description });
    }

    return this.configureBatch({ operations });
  }

  /**
   * Delete a site-to-site peer
   */
  async deletePeer(peerAddress: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "delete_peer", peer: peerAddress }],
    });
  }

  /**
   * Add a tunnel to a peer
   */
  async addPeerTunnel(
    peerAddress: string,
    tunnelId: string,
    options: {
      espGroup?: string;
      localPrefix?: string;
      remotePrefix?: string;
      protocol?: string;
    }
  ): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{
        op: "add_peer_tunnel",
        peer: peerAddress,
        tunnel: tunnelId,
        esp_group: options.espGroup,
        local_prefix: options.localPrefix,
        remote_prefix: options.remotePrefix,
        protocol: options.protocol,
      }],
    });
  }

  /**
   * Delete a tunnel from a peer
   */
  async deletePeerTunnel(peerAddress: string, tunnelId: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "delete_peer_tunnel", peer: peerAddress, tunnel: tunnelId }],
    });
  }

  /**
   * Disable a peer
   */
  async disablePeer(peerAddress: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "disable_peer", peer: peerAddress }],
    });
  }

  /**
   * Enable a peer
   */
  async enablePeer(peerAddress: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "enable_peer", peer: peerAddress }],
    });
  }

  /**
   * Configure VTI for a peer
   */
  async configurePeerVTI(
    peerAddress: string,
    vtiInterface: string,
    espGroup?: string
  ): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{
        op: "set_peer_vti",
        peer: peerAddress,
        bind: vtiInterface,
        esp_group: espGroup,
      }],
    });
  }

  // =========================================================================
  // Interface Helper Methods
  // =========================================================================

  /**
   * Add an IPsec interface
   */
  async addInterface(interfaceName: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "add_ipsec_interface", interface: interfaceName }],
    });
  }

  /**
   * Remove an IPsec interface
   */
  async removeInterface(interfaceName: string): Promise<VyOSResponse> {
    return this.configureBatch({
      operations: [{ op: "delete_ipsec_interface", interface: interfaceName }],
    });
  }

  // =========================================================================
  // Quick Setup Helper
  // =========================================================================

  /**
   * Quick setup for a basic site-to-site VPN
   */
  async quickSetupSiteToSite(config: {
    peerAddress: string;
    localAddress: string;
    preSharedKey: string;
    localNetwork: string;
    remoteNetwork: string;
    ikeGroupName?: string;
    espGroupName?: string;
  }): Promise<VyOSResponse> {
    const ikeGroup = config.ikeGroupName || "IKE-DEFAULT";
    const espGroup = config.espGroupName || "ESP-DEFAULT";

    const operations: IPsecOperation[] = [
      // Create IKE group with modern defaults
      { op: "create_ike_group", name: ikeGroup },
      { op: "set_ike_group_key_exchange", name: ikeGroup, value: "ikev2" },
      { op: "set_ike_group_lifetime", name: ikeGroup, value: 28800 },
      {
        op: "add_ike_group_proposal",
        name: ikeGroup,
        proposal: "1",
        dh_group: "14",
        encryption: "aes256",
        hash: "sha256",
      },
      {
        op: "set_ike_group_dpd",
        name: ikeGroup,
        action: "restart",
        interval: 30,
        timeout: 120,
      },

      // Create ESP group
      { op: "create_esp_group", name: espGroup },
      { op: "set_esp_group_lifetime", name: espGroup, value: 3600 },
      { op: "set_esp_group_pfs", name: espGroup, value: "dh-group14" },
      {
        op: "add_esp_group_proposal",
        name: espGroup,
        proposal: "1",
        encryption: "aes256",
        hash: "sha256",
      },

      // Create peer
      { op: "create_peer", peer: config.peerAddress },
      { op: "set_peer_ike_group", peer: config.peerAddress, value: ikeGroup },
      { op: "set_peer_local_address", peer: config.peerAddress, value: config.localAddress },
      { op: "set_peer_auth_mode", peer: config.peerAddress, value: "pre-shared-secret" },
      { op: "set_peer_auth_psk", peer: config.peerAddress, value: config.preSharedKey },
      { op: "set_peer_connection_type", peer: config.peerAddress, value: "initiate" },

      // Add tunnel
      {
        op: "add_peer_tunnel",
        peer: config.peerAddress,
        tunnel: "1",
        esp_group: espGroup,
        local_prefix: config.localNetwork,
        remote_prefix: config.remoteNetwork,
      },
    ];

    return this.configureBatch({ operations });
  }
}

export const ipsecService = new IPsecService();
